---
title: Analysis of CREMA Gata2 circuitry of the mouse pituitary
author: "Zidong"
date: '`r format(Sys.Date(), "%Y-%B-%d")`'
output:
    html_document:
        toc: true
        toc_depth: 2
        toc_float: true
---

```{r}
# 
# dyn.load("~/anaconda3/lib/libgfortran.so.4")
# dyn.load("~/anaconda3/lib/libpng16.so.16")

library(Seurat)
library(Signac)

library(Matrix)
library(matrixStats)

library(monocle3)
library(cicero)

library(TFBSTools)
library(motifmatchr)

library(EnsDb.Mmusculus.v79)
library(BSgenome.Mmusculus.UCSC.mm10)

library(igraph)

library(ggplot2)
library(pheatmap)
library(RColorBrewer)
library(gplots)

library(VennDiagram)

# library(glmnet)
# library(MASS)

# library(openxlsx)

library(parallel)

library(PRROC)

library(biomaRt)
# library(GENIE3)

# Go term
library(limma)
library(org.Mm.eg.db)

# Lasso
library(glmnet)

# DE
library(DESeq2)
library(apeglm)
library(MASS)
```

```{r}
devtools::load_all("../CREMA/")
```

# Data

## Data

```{r}
sample_id <- "WT22"
dataset_name <- sample_id
```

```{r}
data_multi <- readRDS(file.path("../data_seurat_object/pituitary_wt/",
                                paste0("MM-", sample_id, "_labelTransferred_peaksReCalled.rds")))

Fragments(data_multi[["ATAC"]]) <- NULL
Fragments(data_multi[["ATAC"]]) <- CreateFragmentObject(path = file.path("../data/pituitary_wt/", paste0("MM-", sample_id), "atac_fragments.tsv.gz"),
                                                        cells = colnames(data_multi))

frag_object_inserts <- CreateFragmentObject(path = file.path("../data/pituitary_wt/", paste0("MM-", sample_id), "atac_inserts.tsv.gz"), 
                                            cells = colnames(data_multi))

DefaultAssay(data_multi) <- "SCT"
print(table(data_multi$cell.ident.transferred))
```

```{r}
cells_select <- colnames(data_multi)[data_multi$cell.ident.transferred %in% setdiff(data_multi$cell.ident.transferred, c("Debris", "S+L doublets"))]
data_multi <- subset(data_multi, cells = cells_select)
print(table(data_multi$cell.ident.transferred))
```

```{r}
# # keep standard chromosome
# temp_ranges <- data_multi[["ATAC"]]@ranges
# temp_annotation <- data_multi[["ATAC"]]@annotation
# temp_peak_access_count_mtx <- data_multi[["ATAC"]]@counts
# temp_fragment <- data_multi[["ATAC"]]@fragments
# 
# temp_chromosomes_keep <- c(paste0("chr", seq(1,19)), "chrX", "chrY")
# temp_peak_access_count_mtx <- temp_peak_access_count_mtx[as.character(seqnames(temp_ranges)) %in% temp_chromosomes_keep, , drop = F]
# 
# data_multi[["ATAC"]] <- NULL
# data_multi[["ATAC"]] <- CreateChromatinAssay(counts = temp_peak_access_count_mtx, 
#                                              sep = c("-", "-"),
#                                              genome = "mm10",
#                                              annotation = temp_annotation,
#                                              fragments = temp_fragment)
# 
# 
# DefaultAssay(data_multi) <- "ATAC"
# data_multi <- FindTopFeatures(data_multi, min.cutoff = 5, verbose = F)
# data_multi <- RunTFIDF(data_multi, verbose = F)
# data_multi <- RunSVD(data_multi, verbose = F)
# 
# # build a joint neighbor graph using both assays
# data_multi <- FindMultiModalNeighbors(data_multi,
#                                       reduction.list = list("pca", "lsi"), 
#                                       dims.list = list(1:50, 2:40),
#                                       modality.weight.name = "RNA.weight",
#                                       verbose = T)
# 
# # build a joint UMAP visualization
# data_multi <- RunUMAP(data_multi, nn.name = "weighted.nn", assay = "RNA", verbose = T)
# DimPlot(data_multi, reduction = "umap", group.by = "cell.ident.transferred", label = T, repel = T) + NoLegend()
```


## Gene region

```{r}
# ================
# gene annotation
# ================
print("using gene annotation from EnsDb.Mmusculus.v79")

# extract gene coordinates from Ensembl, and ensure name formatting is consistent with  Seurat object 
# gene.coords <- genes(EnsDb.Mmusculus.v79, filter = ~ gene_biotype == "protein_coding")
gene.coords <- ensembldb::genes(EnsDb.Mmusculus.v79)
# seqlevelsStyle(gene.coords) <- 'UCSC'
# gene.coords <- gene.coords[gene.coords$gene_biotype == "protein_coding"]
genebody.coords <- keepStandardChromosomes(gene.coords, species = "Mus_musculus", pruning.mode = 'coarse')

# add TSS
genebody.coords$TSS <- ifelse(test = (strand(genebody.coords) == "+" | strand(genebody.coords) == "*"),
                              yes = start(genebody.coords),
                              no = end(genebody.coords))

# change style of seqnames
seqlevels(genebody.coords) <- paste0("chr", seqlevels(genebody.coords))
```

```{r}
# for duplicate gene names select "protein_coding"
genebody.coords <- genebody.coords[sapply(genebody.coords$gene_name, nchar) > 0]
genebody.coords_list <- split(genebody.coords, f = genebody.coords$gene_name)
temp_ind <- names(genebody.coords_list)[sapply(genebody.coords_list, length) > 1]

temp_func <- function(x){
    if ("protein_coding" %in% x$gene_biotype){ return(x[x$gene_biotype == "protein_coding"])
    }else{ return(x) }
}
genebody.coords_list <- c(lapply(genebody.coords_list[temp_ind], temp_func),
                          as.list(genebody.coords_list[setdiff(names(genebody.coords_list), temp_ind)]))
genebody.coords <- unlist(as(genebody.coords_list, "GRangesList"))
genebody.coords <- sort(genebody.coords)
```


## Motif match

precomputed on lab desktop (`TFBSTools` can't install on R 3.6.0)

```{r}
motif_database <- "cisbp"
# motif_database <- "jaspar"

motifs <- readRDS(file.path("../resource/tf_motifs/data_rds",
                            paste0("motifs_pwmlist_mouse_", motif_database, ".rds")))

motif_tfs <- unique(sapply(names(motifs), function(x){strsplit(x, split="_")[[1]][1]}))

```

## KO data

### RNA

```{r}
TF_select <- "GATA2"
data_merge <- readRDS(file.path("../data_seurat_object/pituitary_mutation/rna_merged/", paste0("MM", TF_select, "_merged_KO-WT.rds")))

data_merge$condition <- factor(data_merge$condition, levels = c("WT", "KO"))
Idents(data_merge) <- data_merge$condition
```

```{r}
data_merge_subset_list <- sapply(c("Gonadotropes", "Somatotropes", "Lactotropes", "Melanotropes"), 
                                 function(x){subset(data_merge, cells = colnames(data_merge)[data_merge$maintype.ident %in% x])},
                                 USE.NAMES = T, simplify = F)
```

### ATAC

```{r}
TF_select <- "GATA2"
data_atac_merge <- readRDS(file.path("../data_seurat_object/pituitary_mutation/atac_merged/", paste0("MM", TF_select, "_merged_KO-WT.rds")))

data_atac_merge$condition <- factor(data_atac_merge$condition, levels = c("WT", "KO"))
Idents(data_atac_merge) <- data_atac_merge$condition
```

```{r}
frag_KO_inserts_list <- list()
for (i in seq(1, length(Fragments(data_atac_merge)))){
    frag_KO_inserts_list[[i]] <- CreateFragmentObject(path = gsub("fragments.tsv.gz", "atac_inserts.tsv.gz", Fragments(data_atac_merge)[[i]]@path, fixed = T),
                                                      cells = Fragments(data_atac_merge)[[i]]@cells)
}
```

```{r}
data_atac_merge_subset_list <- mclapply(list(Gonadotropes = c("Gonadotropes"), 
                                             Somatotropes = c("Somatotropes", "Somatotropes 1", "Somatotropes 2", "Somatotropes 3"), 
                                             Lactotropes = c("Lactotropes"), 
                                             Melanotropes = c("Melanotropes")),
                                        function(x){subset(data_atac_merge, cells = colnames(data_atac_merge)[data_atac_merge$cell.ident %in% x])})
```


## Pseudo-bulk function

```{r}
pb_DE <- function(data_merge_subset, 
                  genes_select, 
                  assay = "RNA", 
                  sample_name = "orig.ident", group_name = "condition", groups = c("WT", "KO"), 
                  total_count_name = paste0("nCount_", assay),
                  latent_vars = NULL,
                  model = c("DESeq2", "lm", "glm"),
                  norm_mtx_only = F){
    
    metadata_df <- data_merge_subset@meta.data
    
    counts_mtx <- data_merge_subset[[assay]]@counts
    
    sample_list <- split(x = colnames(counts_mtx), f = metadata_df[colnames(counts_mtx), sample_name])
    
    
    # total counts
    total_count <- sapply(sample_list, function(y){sum(metadata_df[y, total_count_name])}, USE.NAMES = T, simplify = T)
    
    # Pseudo-bulk mtx
    sum_mtx <- do.call("cbind", lapply(sample_list, function(y){matrix(rowSums(counts_mtx[, y]), nrow = nrow(counts_mtx), ncol = 1)}))
    row.names(sum_mtx) <- row.names(counts_mtx)
    colnames(sum_mtx) <- names(sample_list)
    
    # normalized pseudo-bulk mtx
    scale_factor <- total_count[1] / median(sum_mtx[,1])
    norm_mtx <- t(t(sum_mtx) / total_count) * scale_factor
    
    if (norm_mtx_only){
        return(norm_mtx)
    }
    
    # total counts mtx
    total_count <- matrix(total_count[colnames(norm_mtx)], nrow = ncol(norm_mtx))
    
    
    # Run DE with the selected model
    model <- match.arg(model)
    
    if (model == "DESeq2"){
        
        # DESeq -----
        
        # sample and group metadata
        sample_meta_df <- metadata_df[, c(sample_name, group_name)]
        sample_meta_df <- sample_meta_df[!duplicated(sample_meta_df), ]
        colnames(sample_meta_df) <- c("sample", "group")
        row.names(sample_meta_df) <- sample_meta_df$sample
        
        sum_mtx <- sum_mtx[, row.names(sample_meta_df)]
        
        # DESeq dataset
        dds <- DESeqDataSetFromMatrix(countData = sum_mtx,
                                      colData = sample_meta_df,
                                      design = ~ group)
        
        # Pre-filter
        dds <- dds[rowSums(counts(dds)) >= 5, ]
    
        # Run DE
        dds <- DESeq(dds)
        
        # Extract results (fold-change: B / A)
        res <- results(dds, 
                       contrast=c("group", groups[2], groups[1]),
                       pAdjustMethod = "fdr")
        resOrdered <- res[order(res$pvalue), ]
        
        # # LFC shrinkage
        # resLFC <- lfcShrink(dds, coef=paste0("group_", groups[2], "_vs_", groups[1]), type="apeglm")
        # resOrdered <- resLFC[order(resLFC$pvalue),]
        
        de_df <- as.data.frame(resOrdered)
        colnames(de_df)[colnames(de_df) == "pvalue"] <- "p_val"
        de_df$padj <- NULL
        de_df$gene <- row.names(de_df)
        de_df$avg_log2FC <- de_df$log2FoldChange
    }
    
    if (model %in% c("lm", "glm", "rlm")){
        
        # group indicator matrix
        sample_group_df <- metadata_df[colnames(counts_mtx), c(sample_name, group_name)]
        sample_group_df <- sample_group_df[!duplicated(sample_group_df), ]
        
        sample_group <- (sample_group_df[, group_name] == groups[2]) + 0
        names(sample_group) <- sample_group_df[, sample_name]
        
        sample_group <- matrix(sample_group[colnames(norm_mtx)], nrow = ncol(norm_mtx))
        
        # # Only test selected genes
        # norm_mtx <- norm_mtx[row.names(norm_mtx) %in% genes_select, , drop = F]
        
        if (model == "lm"){
            # de_lm <- sapply(row.names(norm_mtx), function(y){summary(lm(t(sum_mtx[y, , drop=F]) ~ sample_group + total_count))}, USE.NAMES = T, simplify = F)
            de_lm <- sapply(row.names(norm_mtx), function(y){summary(lm(t(norm_mtx[y, , drop=F]) ~ sample_group))}, USE.NAMES = T, simplify = F)
            de_df <- do.call("rbind", lapply(de_lm, function(y){as.data.frame(coef(y))["sample_group", , drop=F]}))
            colnames(de_df)[colnames(de_df) == "Pr(>|t|)"] <- "p_val"
            de_df$avg_log2FC <- de_df$`Estimate`
        }
        if (model == "glm"){
            de_lm <- sapply(row.names(norm_mtx), function(y){summary(glm(t(sum_mtx[y, , drop=F]) ~ sample_group + total_count, family = poisson(link = "log")))}, USE.NAMES = T, simplify = F)
            # de_lm <- sapply(row.names(norm_mtx), function(y){summary(glm(t(norm_mtx[y, , drop=F]) ~ sample_group, family = poisson(link = "log")))}, USE.NAMES = T, simplify = F)
            de_df <- do.call("rbind", lapply(de_lm, function(y){as.data.frame(coef(y))["sample_group", , drop=F]}))
            colnames(de_df)[colnames(de_df) == "Pr(>|z|)"] <- "p_val"
            de_df$avg_log2FC <- de_df$`Estimate`
        }
        # if (model == "rlm"){
        #     # de_lm <- sapply(row.names(norm_mtx), function(y){summary(rlm(t(sum_mtx[y, , drop=F]) ~ sample_group + total_count))}, USE.NAMES = T, simplify = F)
        #     de_lm <- sapply(row.names(norm_mtx), function(y){summary(rlm(t(norm_mtx[y, , drop=F]) ~ sample_group))}, USE.NAMES = T, simplify = F)
        #     de_df <- do.call("rbind", lapply(de_lm, function(y){as.data.frame(coef(y))["sample_group", , drop=F]}))
        # }
        
        de_df$gene <- names(de_lm)
        row.names(de_df) <- de_df$gene
    }
    
    de_df$p_val_adj <- p.adjust(de_df$p_val, method = "fdr")

    return(de_df[de_df$gene %in% genes_select, , drop = F])
}
```

### Genes DE in KO data

```{r}
pseudobulk_DE_genes_df_list <- mclapply(data_merge_subset_list,
                                        function(x){pb_DE(data_merge_subset = x,
                                                          assay = "SCT",
                                                          # genes_select = names(crema_regions),
                                                          genes_select = row.names(x@assays$SCT@counts),
                                                          model = "DESeq2")},
                                        mc.cores = length(data_merge_subset_list))
for (x in names(pseudobulk_DE_genes_df_list)){
    hist(pseudobulk_DE_genes_df_list[[x]]$p_val, breaks = 30, main = paste(x, sum(pseudobulk_DE_genes_df_list[[x]]$p_val < 5e-3, na.rm = T)))
}
```


## CREMA results

```{r}
# CREMA results =========
motif_database <- "cisbp"

site_extension <- 200
regression_method <- "ols"

smooth_k <- 0
smooth_name <- ifelse(smooth_k == 0, yes = "", no = paste0("smooth-", smooth_k, "_"))

output_dir <- file.path("../linear_model_results/", paste0("pituitary_wt-all-proximal", "-site", site_extension, "_", motif_database))

lm.models_tf_ATACweighted_highres <- readRDS(file.path(output_dir,
                                                       paste0("crema-", regression_method, "-", "SCTcounts",
                                                              "_allTFs-allGenes-ATACweighted-highres_",
                                                              smooth_name,
                                                              dataset_name, ".rds")))
lm.models_tf_ATACweighted_highres <- lm.models_tf_ATACweighted_highres[!is.na(lm.models_tf_ATACweighted_highres)]
lm.models_tf_ATACweighted_highres <- lm.models_tf_ATACweighted_highres[sapply(lm.models_tf_ATACweighted_highres, length) > 1]

temp_p <- unlist(lapply(lm.models_tf_ATACweighted_highres, function(x){x$p[x$t > 0]}))
p_val_cutoff <- max(temp_p[p.adjust(temp_p, method = "fdr") < 1e-4])
print(p_val_cutoff)

temp_func <- function(x){
    temp_p <- lm.models_tf_ATACweighted_highres[[x]]$p
    temp_t <- lm.models_tf_ATACweighted_highres[[x]]$t
    temp_ind <- !is.na(temp_p) & temp_p < p_val_cutoff & !is.na(temp_t) & temp_t > 0
    if(sum(temp_ind) > 0){
        return(data.frame(gene = x,
                          site = names(temp_p)[temp_ind],
                          p_val = temp_p[temp_ind]))
    }else{
        return(data.frame())
    }
}
crema_sites_select_df <- do.call("rbind", mclapply(names(lm.models_tf_ATACweighted_highres), temp_func, mc.cores = 40))
print(nrow(crema_sites_select_df))

temp <- strsplit(crema_sites_select_df$site, split = "_")
crema_sites_select_df$TF <- sapply(temp, function(x){x[1]})
crema_sites_select_df$site <- sapply(temp, function(x){x[2]})
```

```{r}
# # Convert to peak space -----
# min_overlap_for_peak <- 0
# 
# sites_all_str <- unique(crema_sites_select_df$site)
# sites_all_gr <- StringToGRanges(sites_all_str)
# 
# peaks_all_gr <- data_atac_merge@assays$peaks_all@ranges
# peaks_all_str <- GRangesToString(peaks_all_gr)
# 
# temp_overlap <- findOverlaps(sites_all_gr, peaks_all_gr, minoverlap = min_overlap_for_peak)
# temp_overlap_df <- data.frame(site = sites_all_str[temp_overlap@from],
#                               overlapped_peak = peaks_all_str[temp_overlap@to])
# 
# crema_sites_select_df <- merge(crema_sites_select_df, temp_overlap_df, by = "site", all.x = T, sort = F)
# crema_sites_select_df$inpeak <- !is.na(crema_sites_select_df$overlapped_peak)
```

### Sites access in KO data

```{r}
sites_select <- unique(crema_sites_select_df$site)
print(length(sites_select))

# site_extension <- 200
# 
# sites_select_list <- list()
# batchSize <- 5000
# for (i in seq(1, ceiling(length(sites_select)/batchSize))){
#     sites_select_list[[i]] <- sites_select[(batchSize*(i-1)+1) : min(batchSize*i, length(sites_select))]
# }
# 
# sites_access_mtx_list <- mclapply(sites_select_list,
#                                   function(x){FeatureMatrix(fragments = frag_KO_inserts_list,
#                                                             features = Extend(StringToGRanges(x), upstream = site_extension, downstream = site_extension, from.midpoint = T),
#                                                             cells = colnames(data_atac_merge),
#                                                             process_n = 2000000,
#                                                             verbose = F)},
#                                   mc.cores = length(sites_select_list))
# sites_access_mtx <- do.call("rbind", sites_access_mtx_list)
# row.names(sites_access_mtx) <- sites_select
# saveRDS(sites_access_mtx, file.path("../application/pituitary_mutation/site_access_mtx.rds"))
sites_access_mtx <- readRDS(file.path("../application/pituitary_mutation/site_access_mtx.rds"))

# sites_access_mtx <- (sites_access_mtx > 0) + 0

data_sites_merge <- CreateChromatinAssay(counts = sites_access_mtx,
                                        fragments = frag_KO_inserts_list,
                                        annotation = data_atac_merge@assays$peaks_all@annotation)
data_sites_merge <- CreateSeuratObject(counts = data_sites_merge,
                                      assay = "sites",
                                      meta.data = data_atac_merge@meta.data)
data_sites_merge@reductions$umap <- data_atac_merge@reductions$umap
saveRDS(data_sites_merge, file.path("../application/pituitary_mutation/data_sites_seurat.rds"))

data_sites_merge <- readRDS(file.path("../application/pituitary_mutation/data_sites_seurat.rds"))
Idents(data_sites_merge) <- data_sites_merge$condition

data_sites_merge_subset_list <- mclapply(list(Gonadotropes = c("Gonadotropes"), 
                                             Somatotropes = c("Somatotropes", "Somatotropes 1", "Somatotropes 2", "Somatotropes 3"), 
                                             Lactotropes = c("Lactotropes"), 
                                             Melanotropes = c("Melanotropes")),
                                        function(x){subset(data_sites_merge, cells = colnames(data_sites_merge)[data_sites_merge$cell.ident %in% x])})
```

### Sites DA in KO data

```{r}
# pseudobulk_DE_sites_df_list <- mclapply(data_sites_merge_subset_list,
#                                         FUN = pb_DE,
#                                         assay = "sites",
#                                         genes_select = row.names(data_sites_merge),
#                                         total_count_name = "nCount_peaks",
#                                         model = "DESeq2",
#                                         mc.cores = length(data_atac_merge_subset_list))
# saveRDS(pseudobulk_DE_sites_df_list, file.path("../application/pituitary_mutation/DE_sites.rds"))

pseudobulk_DE_sites_df_list <- readRDS(file.path("../application/pituitary_mutation/DE_sites.rds"))
for (x in names(pseudobulk_DE_sites_df_list)){
    hist(pseudobulk_DE_sites_df_list[[x]]$p_val, breaks = 30, main = paste(x, sum(pseudobulk_DE_sites_df_list[[x]]$p_val < 5e-3, na.rm = T)))
}
```


# Gata2 Circuits

## All Gata2 circuits

```{r}
tf_select <- "Gata2"
sites_select_df <-  crema_sites_select_df[crema_sites_select_df$TF == tf_select, ]
sites_select_df <- sites_select_df[order(sites_select_df$p_val), ]

print(length(unique(sites_select_df$site)))
print(length(unique(sites_select_df$gene)))

# Filter for active circuits in each cell type =====

celltypes_mapping <- list(Gonadotropes = c("Gonadotropes"),
                          Somatotropes = c("Somatotropes 1", "Somatotropes 2"))


temp_func <- function(x){
    temp_sites <- unique(sites_select_df$site)
    temp_sites_mtx <- FeatureMatrix(fragments = frag_object_inserts, 
                                    process_n = 200000,
                                    features = Extend(StringToGRanges(temp_sites), upstream = site_extension, downstream = site_extension), 
                                    cells = colnames(data_multi)[data_multi$cell.ident.transferred %in% x], verbose = F)
    row.names(temp_sites_mtx) <- temp_sites
    return(temp_sites_mtx)
}
temp_sites_mtx_list <- mclapply(celltypes_mapping, temp_func, mc.cores = length(celltypes_mapping))
```

```{r}
temp_func <- function(celltype_select){
    
    # filter for active circuits in the selected cell type
    cells_select <- colnames(data_multi)[data_multi$cell.ident.transferred %in% celltypes_mapping[[celltype_select]]]
    
    temp_genes <- unique(sites_select_df$gene)
    temp_genes_mtx <- data_multi[["SCT"]]@counts[c(temp_genes, tf_select), cells_select, drop = F]
    
    temp_sites_mtx <- temp_sites_mtx_list[[celltype_select]]
    
    temp_circuits_mtx <- data_multi[["SCT"]]@counts[sites_select_df$TF, cells_select, drop = F] *
        temp_sites_mtx[sites_select_df$site, cells_select, drop = F] * 
        data_multi[["SCT"]]@counts[sites_select_df$gene, cells_select, drop = F]
    row.names(temp_circuits_mtx) <- paste0(sites_select_df$TF, "_", sites_select_df$site, "_", sites_select_df$gene)
    
    temp_pairs_mtx <- data_multi[["SCT"]]@counts[sites_select_df$TF, cells_select, drop = F] *
        data_multi[["SCT"]]@counts[sites_select_df$gene, cells_select, drop = F]
    row.names(temp_pairs_mtx) <- paste0(sites_select_df$TF, "_", sites_select_df$gene)
    
    
    temp_genes_active <- rowMeans(temp_genes_mtx > 0)
    hist(temp_genes_active, breaks = 30, main = celltype_select[1])
    temp_genes_active <- names(temp_genes_active)[temp_genes_active >= 0.30]
    
    temp_genes_active_byNum <- rowSums(temp_genes_mtx > 0)
    hist(temp_genes_active_byNum, breaks = 30, main = celltype_select[1])
    temp_genes_active_byNum <- names(temp_genes_active_byNum)[temp_genes_active_byNum >= 260]
    
    temp_sites_active <- rowMeans(temp_sites_mtx > 0)
    hist(temp_sites_active, breaks = 30, main = celltype_select[1])
    temp_sites_active <- names(temp_sites_active)[temp_sites_active > 0.03]
    
    temp_sites_active_byNum <- rowSums(temp_sites_mtx > 0)
    hist(temp_sites_active_byNum, breaks = 30, main = celltype_select[1])
    temp_sites_active_byNum <- names(temp_sites_active_byNum)[temp_sites_active_byNum > 25]
    
    temp_circuits_active <- rowMeans(temp_circuits_mtx > 0)
    hist(temp_circuits_active, breaks = 30, main = celltype_select[1])
    temp_circuits_active <- names(temp_circuits_active)[temp_circuits_active > 0.005]
    
    temp_circuits_active_byNum <- rowSums(temp_circuits_mtx > 0)
    hist(temp_circuits_active_byNum, breaks = 30, main = celltype_select[1])
    temp_circuits_active_byNum <- names(temp_circuits_active_byNum)[temp_circuits_active_byNum >= 3]
    
    temp_pairs_active <- rowMeans(temp_pairs_mtx > 0)
    hist(temp_pairs_active, breaks = 30, main = celltype_select[1])
    temp_pairs_active <- names(temp_pairs_active)[temp_pairs_active > 0.1]
    
    temp_pairs_active_byNum <- rowSums(temp_pairs_mtx > 0)
    hist(temp_pairs_active_byNum, breaks = 30, main = celltype_select[1])
    temp_pairs_active_byNum <- names(temp_pairs_active_byNum)[temp_pairs_active_byNum > 100]
    
    return(lapply(list(cells = cells_select,
                       active_genes = temp_genes_active,
                       active_genes_byNum = temp_genes_active_byNum,
                       active_sites = temp_sites_active,
                       active_sites_byNum = temp_sites_active_byNum,
                       active_circuits = temp_circuits_active,
                       active_circuits_byNum = temp_circuits_active_byNum,
                       active_pairs = temp_pairs_active,
                       active_pairs_byNum = temp_pairs_active_byNum),
                  unique))
}

selected_active_circuits_list <- sapply(names(celltypes_mapping), temp_func, USE.NAMES = T, simplify = F)

print(sapply(selected_active_circuits_list, function(x){sapply(x, length)}))
```


```{r}
# select the active genes and sites to analyze on

# # Use the active circuits
# selected_active_sites <- unique(sapply(strsplit(selected_active_circuits_list$Gonadotropes$active_circuits_byNum, split = "_"), function(x){x[2]}))
# selected_active_genes <- unique(sapply(strsplit(selected_active_circuits_list$Gonadotropes$active_circuits_byNum, split = "_"), function(x){x[3]}))
# 
# print(length(selected_active_sites))
# print(length(selected_active_genes))
# 
# selected_active_sites_list <- lapply(selected_active_circuits_list, function(y){unique(sapply(strsplit(y$active_circuits, split = "_"), function(x){x[2]}))})
# selected_active_genes_list <- lapply(selected_active_circuits_list, function(y){unique(sapply(strsplit(y$active_circuits, split = "_"), function(x){x[3]}))})
# 
# print(sapply(selected_active_sites_list, length))
# print(sapply(selected_active_genes_list, length))
```


### All sites

```{r}
# singlecell_DE_sites_df_list <- mclapply(data_sites_merge_subset_list,
#                                         FUN = FindMarkers,
#                                         slot = "counts", 
#                                         features = selected_active_sites, 
#                                         ident.1 = "KO", ident.2 = "WT",
#                                         latent.vars = c("passed_filters"), test.use = "LR",
#                                         logfc.threshold = -1, min.pct = -1, verbose = F,
#                                         mc.cores = length(data_sites_merge_subset_list))
# 
# temp_func <- function(x){
#     x$p_val_adj <- p.adjust(x$p_val, method = "fdr")
#     return(x)
# }
# singlecell_DE_sites_df_list <- lapply(singlecell_DE_sites_df_list, temp_func)
# print(sapply(singlecell_DE_sites_df_list, function(x){x["chr13-75028714-75028724", ]}))
# print(sapply(singlecell_DE_sites_df_list, function(x){sum(x$p_val_adj < 0.05) / nrow(x)}))
```

```{r}
# pseudobulk_DE_sites_df_list <- mclapply(data_sites_merge_subset_list,
#                                         FUN = pb_DE,
#                                         assay = "sites",
#                                         genes_select = selected_active_sites,
#                                         total_count_name = "nCount_peaks",
#                                         model = "DESeq2",
#                                         mc.cores = length(data_atac_merge_subset_list))
# 
# for (x in names(pseudobulk_DE_sites_df_list)){
#     hist(pseudobulk_DE_sites_df_list[[x]]$p_val, breaks = 30, main = paste(x, sum(pseudobulk_DE_sites_df_list[[x]]$p_val < 5e-2, na.rm = T)))
# }
# 
# print(sapply(pseudobulk_DE_sites_df_list, function(x){x["chr13-75028714-75028724", ]}))
```

```{r}
# sites_select_DE_df_list <- lapply(pseudobulk_DE_sites_df_list, function(x){x[x$gene %in% selected_active_sites, ]})
# sites_select_DE_df_list <- lapply(pseudobulk_DE_sites_df_list, function(x){x[x$gene %in% sites_select_df$site, ]})
# 
# sites_select_DE_df_list <- lapply(pseudobulk_DE_sites_df_list, function(x){x[x$gene %in% selected_active_sites_list$Gonadotropes, ]})
# sites_select_DE_df_list <- lapply(pseudobulk_DE_sites_df_list, function(x){x[x$gene %in% selected_active_sites_list$Somatotropes, ]})
# 
# sites_select_DE_df_list <- lapply(pseudobulk_DE_sites_df_list, function(x){x[x$gene %in% selected_active_circuits_list$Gonadotropes$active_sites_byNum, ]})
# sites_select_DE_df_list <- lapply(pseudobulk_DE_sites_df_list, function(x){x[x$gene %in% selected_active_circuits_list$Somatotropes$active_sites_byNum, ]})




for (celltype_select in c("Gonadotropes", "Somatotropes")){
    sites_select_DE_df_list <- lapply(pseudobulk_DE_sites_df_list, 
                                      function(x){x[x$gene %in% selected_active_circuits_list[[celltype_select]]$active_sites_byNum, ]})
    temp_func <- function(x){
        x$p_val_adj <- p.adjust(x$p_val, method = "fdr")
        return(x)
    }
    sites_select_DE_df_list <- lapply(sites_select_DE_df_list, temp_func)
    
    for (x in names(sites_select_DE_df_list)[1:2]){
        hist(sites_select_DE_df_list[[x]]$p_val, breaks = 30, 
             main = paste(x, 
                          sum(sites_select_DE_df_list[[x]]$p_val_adj < 5e-2 & 
                                  sites_select_DE_df_list[[x]]$avg_log2FC < 0, na.rm = T), 
                          "/", nrow(sites_select_DE_df_list[[x]])))
    }
}

# for (celltype_select in names(celltypes_mapping)){
#     sites_select_DE_df_list <- lapply(pseudobulk_DE_sites_df_list, function(x){x[x$gene %in% selected_active_circuits_list[[celltype_select]]$active_sites, ]})
#     temp_func <- function(x){
#         x$p_val_adj <- p.adjust(x$p_val, method = "fdr")
#         return(x)
#     }
#     sites_select_DE_df_list <- lapply(sites_select_DE_df_list, temp_func)
#     
#     for (x in names(sites_select_DE_df_list)[1:2]){
#         hist(sites_select_DE_df_list[[x]]$p_val, breaks = 30, 
#              main = paste(x, sum(sites_select_DE_df_list[[x]]$p_val_adj < 5e-2 & sites_select_DE_df_list[[x]]$avg_log2FC < 0, na.rm = T), "/", nrow(sites_select_DE_df_list[[x]])))
#     }
# }
```


### All targets

```{r}
temp_num_df <- list()

for (celltype_select in c("Gonadotropes", "Somatotropes")){
    
    targets_select_DE_df <- pseudobulk_DE_genes_df_list[[celltype_select]]
    
    targets_select_DE_df_select <- targets_select_DE_df[targets_select_DE_df$gene %in% selected_active_circuits_list[[celltype_select]]$active_genes_byNum, ]
    targets_select_DE_df_select$p_val_adj <- p.adjust(targets_select_DE_df_select$p_val, method = "fdr")
    
    targets_select_DE_df_select_sig <- targets_select_DE_df_select[targets_select_DE_df_select$p_val_adj < 0.05 & 
                                                                       targets_select_DE_df_select$avg_log2FC < 0, ]
    
    hist(targets_select_DE_df_select$p_val, breaks = 30, 
         main = paste(x, nrow(targets_select_DE_df_select_sig), "/", nrow(targets_select_DE_df_select)))
    
    temp_num_total_down_sig <- sum(targets_select_DE_df$p_val < max(targets_select_DE_df_select_sig$p_val, na.rm = T) &
                                       targets_select_DE_df$log2FoldChange < 0, na.rm = T)
    
    temp_num_total_down_sig <- sum(targets_select_DE_df$p_val_adj < 0.05 &
                                       targets_select_DE_df$log2FoldChange < 0, na.rm = T)
    
    temp_p <- phyper(q = nrow(targets_select_DE_df_select_sig),
                     m = temp_num_total_down_sig,
                     n = nrow(targets_select_DE_df) - temp_num_total_down_sig,
                     k = nrow(targets_select_DE_df_select),
                     lower.tail = F)
    
    temp_num_df[[celltype_select]] <- data.frame(celltype = celltype_select,
                                                 num = nrow(targets_select_DE_df_select_sig),
                                                 num_total = nrow(targets_select_DE_df_select),
                                                 p_val = temp_p)
}
temp_num_df <- do.call("rbind", temp_num_df)


temp_num_df <- data.frame(celltype = c("Somatotropes", "Gonadotropes"),
                          num = c(0.05, 10),
                          num_text = c(0,10))

prop_test <- prop.test(x = c(10,0), n = c(88,198), alternative = "greater")

temp_num_df$celltype <- factor(temp_num_df$celltype, levels = c("Somatotropes", "Gonadotropes"))
ggplot(temp_num_df) + 
    geom_bar(aes(x = celltype, y = num, fill = celltype), stat = "identity", width = 0.5) + 
    scale_fill_manual(values = c("#3a5597", "#bf570f")) + 
    scale_y_continuous(breaks = c(0,2,4,6,8,10), position = "right") + #ylim(-0.01, 11) + 
    ylab("Number of down-regulated genes") + 
    # geom_text(aes(x = celltype, y = num + 1, label = num_text), size = 5) +
    coord_flip() + 
    theme_classic() + 
    theme(axis.title.x = element_blank(),
          # axis.text.x = element_text(size = 16, angle = 45, hjust = 1, vjust = 1),
          # axis.text.x = element_blank(),
          # axis.title.y = element_text(size = 16),
          axis.title.y = element_blank(),
          axis.text.y = element_blank(),
          legend.position = "none")
ggsave("../application/pituitary_mutation/Gata2-targets-celltype_num.pdf", width = 2.3, height = 3)
```

```{r}
# targets_select_DE_df_list <- lapply(pseudobulk_DE_genes_df_list, function(x){x[x$gene %in% selected_active_genes, ]})
# 
# targets_select_DE_df_list <- lapply(pseudobulk_DE_genes_df_list, function(x){x[x$gene %in% selected_active_circuits_list$Gonadotropes$active_genes_byNum, ]})
# targets_select_DE_df_list <- lapply(pseudobulk_DE_genes_df_list, function(x){x[x$gene %in% selected_active_circuits_list$Somatotropes$active_genes_byNum, ]})


for (celltype_select in c("Gonadotropes", "Somatotropes")){
    
    targets_select_DE_df_list <- lapply(pseudobulk_DE_genes_df_list, function(x){x[x$gene %in% selected_active_circuits_list[[celltype_select]]$active_genes_byNum, ]})
    temp_func <- function(x){
        x$p_val_adj <- p.adjust(x$p_val, method = "fdr")
        return(x)
    }
    targets_select_DE_df_list <- lapply(targets_select_DE_df_list, temp_func)
    
    for (x in names(targets_select_DE_df_list)[1:2]){
        hist(targets_select_DE_df_list[[x]]$p_val, breaks = 30, 
             main = paste(x, sum(targets_select_DE_df_list[[x]]$p_val_adj < 5e-2 & 
                                     targets_select_DE_df_list[[x]]$avg_log2FC < 0, na.rm = T), 
                          "/", nrow(targets_select_DE_df_list[[x]])))
    }
}


# for (celltype_select in names(celltypes_mapping)){
#     
#     targets_select_DE_df_list <- lapply(pseudobulk_DE_genes_df_list, function(x){x[x$gene %in% selected_active_circuits_list[[celltype_select]]$active_genes, ]})
#     temp_func <- function(x){
#         x$p_val_adj <- p.adjust(x$p_val, method = "fdr")
#         return(x)
#     }
#     targets_select_DE_df_list <- lapply(targets_select_DE_df_list, temp_func)
#     
#     for (x in names(targets_select_DE_df_list)[1:2]){
#         hist(targets_select_DE_df_list[[x]]$p_val, breaks = 30, 
#              main = paste(x, sum(targets_select_DE_df_list[[x]]$p_val_adj < 0.1 & targets_select_DE_df_list[[x]]$avg_log2FC < 0, na.rm = T), "/", nrow(targets_select_DE_df_list[[x]])))
#     }
# }
```


## Gata2-Pcsk1 circuit

```{r}
tf_select <- "Gata2"
target_select <- "Pcsk1"

temp_df <- crema_sites_select_df[crema_sites_select_df$gene == target_select & crema_sites_select_df$TF == tf_select, ]
temp_df <- temp_df[order(temp_df$p_val), ]
print(temp_df)
```

```{r}
# Get the ATAC counts in the extended site range
temp_gr_str <- unique(temp_df$site)
temp_gr <- Extend_Granges(StringToGRanges(temp_gr_str), upstream = site_extension, downstream = site_extension, from.midpoint = T)

sites_count_mtx <- FeatureMatrix(fragments = Fragments(data_multi[["ATAC"]]),
                                 features = temp_gr, 
                                 process_n = 2000000, 
                                 cells = colnames(data_multi))
row.names(sites_count_mtx) <- temp_gr_str
sites_count_mtx <- sites_count_mtx > 0 + 0

data_multi_temp <- CreateChromatinAssay(counts = sites_count_mtx,
                                        fragments = Fragments(data_multi[["ATAC"]]),
                                        annotation = data_multi[["ATAC"]]@annotation)
data_multi_temp <- CreateSeuratObject(counts = data_multi_temp,
                                      assay = "ATAC",  project = "ATAC",
                                      meta.data = data_multi@meta.data)
data_multi_temp@reductions$umap <- data_multi@reductions$umap
```

```{r}
tf_select <- "Gata2"
target_select <- "Pcsk1"

out_dir <- file.path("../application/pituitary_wt/", paste0("Examples_", tf_select, "-", target_select))
dir.create(out_dir, showWarnings = F, recursive = T)

fig_print <- T
fig_save <- F

sites_select_df

temp_target <- target_select
temp_tf <- tf_select
temp_site <- temp_df$site[1]


# Scatter plot -----
p1 <- FeaturePlot(data_multi, features = c(temp_target, temp_tf), reduction = "umap", pt.size = 0.1)
p2 <- FeaturePlot(data_multi_temp, features = c(temp_site), reduction = "umap", pt.size = 0.1)
if (fig_print){ print(p1 + p2) }
if (fig_save){
    ggsave(file.path(out_dir, paste0("scatter_", temp_tf, "-", temp_target, "-", temp_site, ".png")),
           p1 + p2, width = 8, height = 8)
}


# Coverage plot -----

temp_site_gr <- Extend(temp_site, upstream = 50, downstream = 50, from.midpoint = T)

plot_gr <- genebody.coords[(genebody.coords$symbol == temp_target)][1]
# plot_gr <- GRanges(seqnames = seqnames(crema_regions[[temp_target]][1]),
#                    ranges = IRanges(start = crema_regions[[temp_target]][1]$TSS, end = crema_regions[[temp_target]][1]$TSS+1))
plot_gr <- GRanges(seqnames = seqnames(plot_gr),
                   ranges = IRanges(start = min(start(plot_gr), start(temp_site_gr)),
                                    end = max(end(plot_gr), end(temp_site_gr))))
plot_gr <- Extend(plot_gr, upstream = 5000, downstream = 5000)

temp_site_gr <- Extend(StringToGRanges(temp_site), upstream = 50, downstream = 50, from.midpoint = T)

# Separate by target gene expression
data_multi$temp_ident <- ifelse(data_multi@assays$SCT@counts[target_select, colnames(data_multi)] > 0,
                                yes = "high", no = "low")
covplot <- CoveragePlot(data_multi, 
                        assay = "ATAC",
                        region = plot_gr,
                        group.by = "temp_ident",
                        annotation = T, peaks = T,
                        ranges = temp_site_gr, ranges.title = temp_tf, 
                        features = c(temp_target, temp_tf))
if (fig_print){ print(covplot) }
if (fig_save){
    ggsave(file.path(out_dir,
                     paste0("covplot-groupExp_", temp_tf, "-", temp_target, "-", temp_site, ".pdf")),
           covplot, width = 6, height = 6)
}

# Separate by site accessibility
data_multi$temp_ident <- ifelse(data_multi_temp@assays$ATAC@counts[site_select, colnames(data_multi)] > 0,
                                yes = "high", no = "low")
covplot <- CoveragePlot(data_multi,
                        assay = "ATAC",
                        region = plot_gr,
                        group.by = "temp_ident",
                        annotation = T, peaks = T,
                        ranges = temp_site_gr, ranges.title = temp_tf,
                        features = c(temp_target, temp_tf))
if (fig_print){ print(covplot) }
if (fig_save){
    ggsave(file.path(out_dir,
                     paste0("covplot-groupSite_", temp_tf, "-", temp_target, "-", temp_site, ".pdf")),
           covplot, width = 6, height = 6)
}


# Zoomed coverage plot ----------

plot_gr <- Extend(StringToGRanges(site_select), upstream = 300, downstream = 300)

data_multi$temp_ident <- ifelse(data_multi@assays$SCT@counts[target_select, colnames(data_multi)] > 0,
                                yes = "high", no = "low")
covplot <- CoveragePlot(data_multi, 
                        assay = "ATAC",
                        region = plot_gr,
                        group.by = "temp_ident",
                        annotation = F, peaks = F)
if (fig_print){ print(covplot) }
if (fig_save){
    ggsave(file.path(out_dir,
                     paste0("covplot-groupExp_", temp_tf, "-", temp_target, "-", temp_site, "_zoom.pdf")),
           covplot, width = 6, height = 6)
}
```


## Pcsk1 in KO data

```{r}
target_select <- "Pcsk1"
print(crema_sites_select_df[crema_sites_select_df$TF == tf_select & crema_sites_select_df$gene == target_select, ])

site_select <- "chr13-75028714-75028724"

targets_select_DE_df_list <- lapply(pseudobulk_DE_genes_df_list, function(x){x[x$gene %in% selected_active_circuits_list$Gonadotropes$active_genes_byNum, ]})
sites_select_DE_df_list <- lapply(pseudobulk_DE_sites_df_list, function(x){x[x$gene %in% selected_active_circuits_list$Gonadotropes$active_sites_byNum, ]})

print(sapply(pseudobulk_DE_genes_df_list, function(x){x[target_select, ]}))
print(sapply(pseudobulk_DE_sites_df_list, function(x){x[site_select, ]}))
```

### RNA

```{r}
# Plot of Pseudobulk DE

tf_select <- "Gata2"
temp_target <- "Pcsk1"

print(sapply(pseudobulk_DE_genes_df_list, function(x){x["Pcsk1", ]}))

target_pbde_mtx <- list()
for (celltypes_select in c("Gonadotropes", "Somatotropes", "Lactotropes", "Melanotropes")){
    
    temp_df <- pb_DE(data_merge_subset_list[[celltypes_select]], genes_select = c(temp_target, tf_select), assay = "RNA", norm_mtx_only = T)
    temp_df <- as.data.frame(temp_df[temp_target, , drop = F])
    
    temp_df$gene <- row.names(temp_df)
    temp_df$celltype <- celltypes_select
    target_pbde_mtx[[celltypes_select]] <- temp_df
}
target_pbde_mtx <- do.call("rbind", target_pbde_mtx)

target_pbde_mtx <- reshape2::melt(target_pbde_mtx, 
                                  id.vars = c("gene", "celltype"), 
                                  variable.name = "sample", value.name = "pseudo_bulk_expression")
temp_sample_group <- data_merge@meta.data[, c("orig.ident", "condition")]
temp_sample_group <- temp_sample_group[!duplicated(temp_sample_group), ]
target_pbde_mtx <- merge(target_pbde_mtx, temp_sample_group, by.x = "sample", by.y = "orig.ident", all.x = T)

ggplot(target_pbde_mtx) + 
    geom_boxplot(aes(x = celltype, y = pseudo_bulk_expression, fill = condition)) +
    scale_fill_manual(values = c("#468499", "#EEAA88")) + 
    ylab("Pseudo-bulk expression") + 
    ggtitle(paste(temp_target, "pseudo-bulk expression")) + 
    theme_minimal() + 
    theme(axis.text.x = element_text(size = 16, angle = 45, hjust = 1, vjust = 1),
          axis.title.x = element_blank(),
          axis.title.y = element_text(size = 16),
          legend.title = element_text(size = 16), legend.text = element_text(size = 16))
    

ggplot(target_pbde_mtx) + 
    geom_point(aes(x = celltype, y = pseudo_bulk_expression, color = condition), position = position_jitterdodge(), size = 3) + 
    scale_color_manual(values = c("#468499", "#EEAA88")) + 
    ylab("Pseudo-bulk expression") + 
    ggtitle(paste(temp_target, "pseudo-bulk expression")) + 
    theme_minimal() + 
    theme(axis.text.x = element_text(size = 16, angle = 45, hjust = 1, vjust = 1),
          axis.title.x = element_blank(),
          axis.title.y = element_text(size = 16),
          legend.title = element_text(size = 16), legend.text = element_text(size = 16))



ggplot(target_pbde_mtx) + 
    geom_boxplot(aes(x = celltype, y = pseudo_bulk_expression, fill = condition), outlier.shape = NA) +
    geom_point(aes(x = celltype, y = pseudo_bulk_expression, fill = condition), position = position_jitterdodge(), size = 3) + 
    scale_fill_manual(values = c("#468499", "#EEAA88"), labels = c("WT", "Gonadotrope KO")) + 
    ylab("Pseudo-bulk expression") + 
    ggtitle(paste(temp_target, "pseudo-bulk expression")) + 
    theme_minimal() + 
    theme(axis.text.x = element_text(size = 16, angle = 45, hjust = 1, vjust = 1),
          axis.title.x = element_blank(),
          axis.title.y = element_text(size = 16),
          legend.title = element_text(size = 16), legend.text = element_text(size = 16))
```

```{r}
# Plot of single cell DE

for (celltypes_select in c("Gonadotropes", "Somatotropes", "Lactotropes", "Melanotropes")){

    data_merge_subset <- data_merge_subset_list[[celltypes_select]]
    
    # p1 <- VlnPlot(data_merge_subset, 
    #               features = c(temp_target, "nCount_SCT"), 
    #               group.by = "condition", 
    #               split.by = "orig.ident", 
    #               cols = rep(c("#468499", "#EEAA88"), each = 3))
    p1 <- VlnPlot(data_merge_subset, 
                  assay = "RNA", slot = "data", 
                  features = c(temp_target), 
                  group.by = "condition", 
                  cols = c("#468499")) +
        ggtitle(paste(temp_target, celltypes_select, temp_df[temp_target, "avg_log2FC"]))
    print(p1)
}


data_merge_subset <- subset(data_merge, cells = colnames(data_merge)[(data_merge$maintype.ident %in% c("Gonadotropes", "Somatotropes", "Lactotropes", "Melanotropes"))])

p1 <- VlnPlot(data_merge_subset, 
              assay = "RNA", slot = "data", 
              features = c(temp_target), 
              group.by = "maintype.ident",
              split.by = "condition",
              cols = c("#468499", "#EEAA88")) +
    ggtitle(paste(temp_target, celltypes_select, temp_df[temp_target, "avg_log2FC"]))
print(p1)
```

### ATAC

```{r}
# Plot of Pseudobulk DE

temp_target <- "Pcsk1"
temp_site <- "chr13-75028714-75028724"

print(sapply(pseudobulk_DE_sites_df_list, function(x){x[temp_site, ]}))

target_pbde_mtx <- list()
for (celltypes_select in c("Gonadotropes", "Somatotropes", "Lactotropes", "Melanotropes")){
    
    temp_df <- pb_DE(data_sites_merge_subset_list[[celltypes_select]], genes_select = temp_site, assay = "sites", total_count_name = "passed_filters", norm_mtx_only = T)
    temp_df <- as.data.frame(temp_df[temp_site, , drop = F])
    
    temp_df$gene <- row.names(temp_df)
    temp_df$celltype <- celltypes_select
    target_pbde_mtx[[celltypes_select]] <- temp_df
}
target_pbde_mtx <- do.call("rbind", target_pbde_mtx)

target_pbde_mtx <- reshape2::melt(target_pbde_mtx, 
                                  id.vars = c("gene", "celltype"), 
                                  variable.name = "sample", value.name = "pseudo_bulk_accessibility")
temp_sample_group <- data_atac_merge@meta.data[, c("orig.ident", "condition")]
temp_sample_group <- temp_sample_group[!duplicated(temp_sample_group), ]
target_pbde_mtx <- merge(target_pbde_mtx, temp_sample_group, by.x = "sample", by.y = "orig.ident", all.x = T)

ggplot(target_pbde_mtx) + 
    geom_boxplot(aes(x = celltype, y = pseudo_bulk_accessibility, fill = condition)) +
    scale_fill_manual(values = c("#468499", "#EEAA88")) + 
    ylab("Pseudo-bulk accessibility") + 
    ggtitle(paste(tf_select, temp_target, "regulatory domain pseudo-bulk accessibility")) + 
    theme_minimal() + 
    theme(axis.text.x = element_text(size = 16, angle = 45, hjust = 1, vjust = 1),
          axis.title.x = element_blank(),
          axis.title.y = element_text(size = 16))
    

ggplot(target_pbde_mtx) + 
    geom_point(aes(x = celltype, y = pseudo_bulk_accessibility, color = condition), position = position_jitterdodge(), size = 3) + 
    scale_color_manual(values = c("#468499", "#EEAA88")) + 
    ylab("Pseudo-bulk accessibility") + 
    ggtitle(paste(tf_select, temp_target, "regulatory domain pseudo-bulk accessibility")) + 
    theme_minimal() + 
    theme(axis.text.x = element_text(size = 16, angle = 45, hjust = 1, vjust = 1),
          axis.title.x = element_blank(),
          axis.title.y = element_text(size = 16))



ggplot(target_pbde_mtx) + 
    geom_boxplot(aes(x = celltype, y = pseudo_bulk_accessibility, fill = condition), outlier.shape = NA) +
    geom_point(aes(x = celltype, y = pseudo_bulk_accessibility, fill = condition), position = position_jitterdodge(), size = 3) + 
    scale_fill_manual(values = c("#468499", "#EEAA88"), labels = c("WT", "Gonadotrope KO")) + 
    ylab("Pseudo-bulk accessibility") + 
    ggtitle(paste(tf_select, temp_target, "regulatory domain pseudo-bulk accessibility")) + 
    theme_minimal() + 
    theme(axis.text.x = element_text(size = 16, angle = 45, hjust = 1, vjust = 1),
          axis.title.x = element_blank(),
          axis.title.y = element_text(size = 16),
          legend.title = element_text(size = 16), legend.text = element_text(size = 16))
```


```{r}
# Plot of single cell DE

for (celltypes_select in names(data_atac_merge_subset_list)){
    
    data_merge_subset <- data_atac_merge_subset_list[[celltypes_select]]
    Idents(data_merge_subset) <- data_merge_subset$condition

    p1 <- VlnPlot(data_merge_subset, features = c(temp_site, "passed_filters"), group.by = "condition", split.by = "orig.ident", cols = rep(c("#468499", "#EEAA88"), each = 3))
    # p1 <- VlnPlot(data_merge_subset, features = c(temp_site), group.by = "condition") + 
    #     ggtitle(paste(temp_target, celltypes_select, temp_df[temp_target, "avg_log2FC"]))
    print(p1)
}
```


```{r}
# Plot of ATAC tracks

cells.keep <- colnames(data_atac_merge_temp)[(data_atac_merge_temp$cell.ident %in% c("Gonadotropes"))]
data_atac_merge_subset <- subset(data_atac_merge_temp, cells = cells.keep)

temp_tf <- tf_select
temp_target <- "Pcsk1"
temp_site <- sites_select_df$site[sites_select_df$gene == temp_target][1]


fig_print <- T
fig_save <- T

# Coverage plot -----
    
temp_site_gr <- Extend(StringToGRanges(temp_site), upstream = 50, downstream = 50, from.midpoint = T)

plot_gr <- GRanges(seqnames = seqnames(crema_regions[[temp_target]][1]),
                   ranges = IRanges(start = crema_regions[[temp_target]][1]$TSS, 
                                    end = crema_regions[[temp_target]][1]$TSS+1))
plot_gr <- GRanges(seqnames = seqnames(plot_gr),
                   ranges = IRanges(start = min(start(plot_gr), start(temp_site_gr)),
                                    end = max(end(plot_gr), end(temp_site_gr))))
plot_gr <- Extend(plot_gr, upstream = 5000, downstream = 5000)

tracks_separated_by <- "orig.ident"

scale_factors <- sapply(split(data_atac_merge_subset$passed_filters, data_atac_merge_subset@meta.data[tracks_separated_by]), median)

covplot <- CoveragePlot(data_atac_merge_subset, 
                        region = plot_gr,
                        group.by = tracks_separated_by,
                        # tile = T,
                        scale.factor = scale_factors,
                        window = 400,
                        annotation = F, 
                        peaks = F,
                        ranges = temp_site_gr, ranges.title = temp_tf)
if (fig_print){ print(covplot) }
if (fig_save){
    ggsave(file.path(out_dir,
                     paste0("covplot_", temp_site[1], "zoom.pdf")),
           covplot, width = 6, height = 6)
}


temp_site_gr <- StringToGRanges(temp_site)
plot_gr <- temp_site_gr
plot_gr <- Extend(plot_gr, upstream = 1000, downstream = 1000)

covplot <- CoveragePlot(data_atac_merge_subset, 
                        region = plot_gr,
                        group.by = tracks_separated_by,
                        # tile = T,
                        scale.factor = scale_factors,
                        window = 100,
                        annotation = F, 
                        peaks = F,
                        ranges = temp_site_gr, ranges.title = temp_tf)
if (fig_print){ print(covplot) }
if (fig_save){
    ggsave(file.path(out_dir,
                     paste0("covplot_", temp_site[1], "_zoom.pdf")),
           covplot, width = 6, height = 6)
}


```




